implemented LedgerInfoWithSignatures.verifySignatures

haroldcarr opened the pull request:
This implements `LedgerInfoWithSignatures.verifySignatures`

That required implementing
- various validation functions in `ValidatorVerifier`
- creating a `CryptoHash` typeclass with `sign` and `verify` functions in `LibraBFT.Impl.OBM.Crypto`
  - implementing a `CryptoHash LedgerInfo` instance
  - other instances will be created when needed

This is work based on the `mark-verifyqcm` branch.

cwjnkins commented:
This looks good to me -- perhaps it would be worthwhile to bring these changes into `main` as well?

haroldcarr commented:
I have updated this to include
- define and use `forM_`
- made `fakeErr` a real data type (to be renamed `ErrLog` in a future pull request)
- use the new `fakeErr` constructor in `ValidatorVerifier` and `PendingVotes`
- removed unnecessary identifiers when signaling instances of `CryptoHash`
- replaced an `if true` stub with the real call
- defined and used `foldrM`

mark-moir reviewed (approved):
@haroldcarr looks good, thanks.  I've made some review comments, but I am going to merge this and then address them myself in a subsequent merge request.
